src/services/pasteEdits.ts

```diff
@@ -104,12 +107,45 @@ function pasteEdits(
                 preferences,
                 formatContext,
             };
-            forEachChild(updatedFile, function cb(node) {
-                if (isIdentifier(node) && !originalProgram?.getTypeChecker().resolveName(node.text, node, SymbolFlags.All, /*excludeGlobals*/ false)) {
-                    // generate imports
-                    importAdder.addImportForUnresolvedIdentifier(context, node, /*useAutoImportProvider*/ true);
-                }
-                node.forEachChild(cb);
+
+            // `updatedRanges` represent the new ranges that account for the offset changes caused by pasting new text and
+            // `offset` represents by how much the starting position of `pasteLocations` needs to be changed.
+            //
+            // We iterate over each updated range to get the node that wholly encloses the updated range. For each child of that node, it is checked if the
+            // identifier lies within the updated range and if it is not resolved, we try resolving it.
+            let offset = 0;
+            pasteLocations.forEach((location, i) => {
+                const oldTextLength = location.end - location.pos;
+                const textToBePasted = actualPastedText ? actualPastedText[0] : pastedText[i];

```

Comment:
Is it ever possible that the count of paste locations exceeds the number of items in `pastedText`?

Reply:
This is done at top of the function to make sure its always right

Reply:
Yes, in that case all `pastedText` is joined together with a newline character and that is pasted at all locations. Line 64 does this.

Reply:
Gotcha - while you're here, can you fix that so it uses the appropriate newline character instead of `\n`?

Reply:
Also, `actualPastedText` doesn't have to be an array if you only ever use it for `actualPastedText[0]`. Just use `actualPastedText ?? pastedText[i]`.

Reply:
Do you mean using `getNewLineOrDefaultFromHost()` for the appropriate newline character?